### PR TITLE
Request keeper

### LINKED ISSUE #495: Implement keeper and type for data requests and data reports
Design and implement keepers and data structures for data requests and reports. 

For data requests, it should store:
- The oracle script identifier
- The parameters of the request
- The list of validators that is responsible for responding to the request
- The list of validators that already submit report transaction
- The block height and block time that this request is prepared
- The block height that this request will expire
- Whether it has been resolved

For data reports, it should store (keyed by request id, external id, validator):
- The data from reporter

### PR DESCRIPTION
**Pull request by taobun**
fixed #495 

A lot of code in query part is commented. Because new request structure

### CODE REVIEW

File: chain/x/zoracle/internal/keeper/request.go [hunk 1]
```diff
@@ -24,16 +24,53 @@ func (k Keeper) GetRequest(ctx sdk.Context, id uint64) (types.Request, sdk.Error
 	return request, nil
 }
 
+func (k Keeper) AddSubmitValidator(ctx sdk.Context, id int64, validator sdk.ValAddress) sdk.Error {
```
Comment (evilpeach) [inline chain/x/zoracle/internal/keeper/request.go]: add comment

File: chain/x/zoracle/internal/keeper/request.go [hunk 2]
```diff
@@ -24,16 +24,53 @@ func (k Keeper) GetRequest(ctx sdk.Context, id uint64) (types.Request, sdk.Error
...
+
+	if !found {
+		return types.ErrInvalidValidator(types.DefaultCodespace)
+	}
+	request.SubmittedValidatorList = append(request.SubmittedValidatorList, validator)
+	k.SetRequest(ctx, id, request)
+	return nil
+}
+
+func (k Keeper) SetResolve(ctx sdk.Context, id int64, isResolved bool) sdk.Error {
```
Comment (evilpeach) [inline chain/x/zoracle/internal/keeper/request.go:56]: ditto

File: chain/x/zoracle/internal/types/request.go [hunk 3]
```diff
@@ -1,32 +1,39 @@
...
-	Params      []byte `json:"params"`
-	ReportEndAt uint64 `json:"reportEnd"`
+	OracleScriptID           int64            `json:"oracleScriptID"`
+	Calldata                 []byte           `json:"calldata"`
+	ValidatorList            []sdk.ValAddress `json:"validatorList"`
+	SufficientValidatorCount int64            `json:"sufficientValidatorCount"`
+	SubmittedValidatorList   []sdk.ValAddress `json:"submittedValidatorList"`
+	RequestedBlock           int64            `json:"requestedBlock"`
+	RequestedTime            int64            `json:"requestedTime"`
+	ExpiredBlock             int64            `json:"expiredBlock"`
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: Let's call there fields `ExpirationHeight` and `RequestHeight`(we use "request" here as an [Adjectival noun](https://en.wikipedia.org/wiki/Noun_adjunct) using "requested height" would mean the "height" is requested - that's not right).

File: chain/x/zoracle/internal/types/request.go [hunk 4]
```diff
@@ -1,32 +1,39 @@
...
 
-// Request is a type to store detail of request
+// Request is a type to store detail of request.
 type Request struct {
-	CodeHash    []byte `json:"codeHash"`
-	Params      []byte `json:"params"`
-	ReportEndAt uint64 `json:"reportEnd"`
+	OracleScriptID           int64            `json:"oracleScriptID"`
+	Calldata                 []byte           `json:"calldata"`
+	ValidatorList            []sdk.ValAddress `json:"validatorList"`
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: Two things.
1. For `ValidatorList`, let's just call it `Validators` or `ActiveValidators`. Using word `List` when the type already tells you it's a list is bit redundant.
2. Do we need `SubmittedValidatorList`? I think we can just iterate on all reports, right? Maybe we just need `ReportCount`?

File: chain/x/zoracle/internal/types/request.go [hunk 5]
```diff
@@ -1,32 +1,39 @@
 package types
 
 import (
-	"fmt"
-	"strings"
+	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-// Request is a type to store detail of request
+// Request is a type to store detail of request.
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: Request is a data structure that stores the detail of a request to an oracle script.

File: chain/x/zoracle/internal/keeper/request_test.go [hunk 6]
```diff
@@ -4,49 +4,150 @@ import (
...
+func TestAddInvalidSubmitValidator(t *testing.T) {
+	ctx, keeper := CreateTestInput(t, false)
+	request := newDefaultRequest()
 
 	keeper.SetRequest(ctx, 1, request)
+
+	err := keeper.AddSubmitValidator(ctx, 1, sdk.ValAddress([]byte("validator3")))
+	require.Equal(t, types.CodeInvalidValidator, err.Code())
+
 	actualRequest, err := keeper.GetRequest(ctx, 1)
```
Comment (evilpeach) [inline chain/x/zoracle/internal/keeper/request_test.go:73]: check `validator3` is not in `actualRequest` ?

File: chain/x/zoracle/internal/types/report.go [hunk 7]
```diff
@@ -53,3 +31,23 @@ func NewExternalData(externalDataID int64, data []byte) ExternalData {
 		Data:           data,
 	}
 }
+
+// ValidatorReport is a report that contain operator address in struct
+type ValidatorReport struct {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/report.go]: Let's call this `ReportWithValidator` for extra clarity.

File: chain/x/zoracle/internal/keeper/request.go (hunk 2 above)
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/request.go:56]: Feels really bad to load the whole struct just to set one boolean value.... But I think this is the standard. So ok.

File: chain/x/zoracle/internal/keeper/report.go [hunk 8]
```diff
@@ -3,27 +3,31 @@ package keeper
...
-func (k Keeper) GetReportsIterator(ctx sdk.Context, requestID uint64) sdk.Iterator {
+func (k Keeper) GetReportsIterator(ctx sdk.Context, requestID int64) sdk.Iterator {
 	prefix := types.GetIteratorPrefix(types.ReportKeyPrefix, requestID)
 	store := ctx.KVStore(k.storeKey)
 	return sdk.KVStorePrefixIterator(store, prefix)
 }
 
 // GetDataReports returns all the reports for a specific request ID.
-func (k Keeper) GetDataReports(ctx sdk.Context, requestID uint64) []types.Report {
+func (k Keeper) GetDataReports(ctx sdk.Context, requestID int64) []types.Report {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: Some of these might not be needed anymore. We will see after we finish the handler (:

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/request_test.go (hunk 6 above)
Comment (taobun) [reply chain/x/zoracle/internal/keeper/request_test.go:73]: Test equal in line 75 already

File: chain/x/zoracle/internal/types/request.go [hunk 9]
```diff
@@ -1,32 +1,39 @@
...
-	ReportEndAt uint64 `json:"reportEnd"`
+	OracleScriptID           int64            `json:"oracleScriptID"`
+	Calldata                 []byte           `json:"calldata"`
+	ValidatorList            []sdk.ValAddress `json:"validatorList"`
+	SufficientValidatorCount int64            `json:"sufficientValidatorCount"`
+	SubmittedValidatorList   []sdk.ValAddress `json:"submittedValidatorList"`
+	RequestedBlock           int64            `json:"requestedBlock"`
+	RequestedTime            int64            `json:"requestedTime"`
+	ExpiredBlock             int64            `json:"expiredBlock"`
+	IsResolved               bool             `json:"isResolved"`
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go:18]: As we already talked, I think we'll need to have a way to track the list of external data requests. Each will have the external ID, the data source ID, and calldata.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/types/request.go (hunk 4 above)
Comment (sorawit) [reply chain/x/zoracle/internal/types/request.go]: Ok so we decide to keep both lists. Let's name it `RequestedValidators` and `ReceivedValidators`.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/raw_request.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
+package keeper
+
+import (
+	"github.com/bandprotocol/d3n/chain/x/zoracle/internal/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// SetRawDataRequest is a function to save raw data request detail to the given request id and external id.
+func (k Keeper) SetRawDataRequest(ctx sdk.Context, requestID, externalID, scriptID int64, calldata []byte) {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/raw_request.go]: shouldn't this be data source id, not script id?

File: chain/x/zoracle/internal/types/key.go [hunk 11]
```diff
@@ -28,48 +28,57 @@ var (
...
+
+	// RawDataReportStoreKeyPrefix is a prefix for report store
+	RawDataReportStoreKeyPrefix = []byte{0x03}
 )
 
 // RequestStoreKey is a function to generate key for each request in store
-func RequestStoreKey(requestID uint64) []byte {
-	buf := uint64ToBytes(requestID)
+func RequestStoreKey(requestID int64) []byte {
+	buf := int64ToBytes(requestID)
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/key.go]: Make it fabulous

File: chain/x/zoracle/internal/types/key.go [hunk 12]
```diff
@@ -28,48 +28,57 @@ var (
...
 	return append(RequestStoreKeyPrefix, buf...)
 }
 
 // ResultStoreKey is a function to generate key for each result in store
-func ResultStoreKey(requestID uint64, codeHash []byte, params []byte) []byte {
-	buf := uint64ToBytes(requestID)
+func ResultStoreKey(requestID int64, codeHash []byte, params []byte) []byte {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/key.go:50]: Let's make it more readable and follow the same pattern.

File: chain/x/zoracle/internal/types/key.go [hunk 13]
```diff
@@ -28,48 +28,57 @@ var (
...
+	buf := append(RawDataReportStoreKeyPrefix, int64ToBytes(requestID)...)
+	buf = append(buf, int64ToBytes(externalID)...)
 	return append(buf, validatorAddress.Bytes()...)
 }
 
 // GetIteratorPrefix is a function to get specific prefix
-func GetIteratorPrefix(prefix []byte, requestID uint64) []byte {
-	buf := uint64ToBytes(requestID)
+func GetIteratorPrefix(prefix []byte, requestID int64) []byte {
+	buf := int64ToBytes(requestID)
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/key.go]: Make it fabulous.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/types/request.go [hunk 14]
```diff
@@ -1,32 +1,54 @@
...
-	return strings.TrimSpace(fmt.Sprintf(`CodeHash: %x
-Params: %x
-ReportEndAt: %d`,
-		req.CodeHash,
-		req.Params,
-		req.ReportEndAt,
-	))
+// RawDataRequest is a data structure that store what datasource and calldata will be used in request.
+type RawDataRequest struct {
+	ScriptID int64  `json:"scriptID"`
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/request.go]: DataSourceID. Not ScriptID.

**Review by sorawit — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**